lib/src/facts/aix/operating_system_resolver.cc

```diff
@@ -1,43 +1,72 @@
 
 #include <boost/algorithm/string.hpp>
+#include <odmi.h>
+#include <sys/cfgodm.h>
 
 using namespace std;
-using namespace leatherman::util;
-using namespace leatherman::file_util;
-using namespace boost;
-namespace execution = leatherman::execution;
+using namespace facter::util::aix;
 
+// This routine's meant to be a general utility function that replicates the behavior of
+// lsattr -El <object> -a <field>. Although it's only used to get the modelname of the
+// sys0 device, we still would like to have it here in case we ever need to separate it
+// out to an AIX utils file to query other attributes. Part of what lsattr does is check
+// PdAt if we don't have an entry for the object's attribute in CdAt, even though it is very
+// unlikely that sys0.modelname will not have a CuAt entry. That is why we have the extra code
+// in here.
 static string getattr(string object, string field)
 {
-    string result;
-
-    execution::each_line(
-        "/usr/sbin/lsattr", {"-El", object, "-a", field},
-        [&](string& line) {
-            if (!line.empty()) {
-                vector<string> tokens;
-                boost::split(tokens, line, boost::is_space(), boost::token_compress_on);
-                if (tokens.size() < 2) {
-                    return true;
-                }
-                result = tokens[1];
-                return false;
-            }
-            return true;
-        },
-        nullptr,
-        0);
-
-    if (result == "") {
-        LOG_WARNING("Could not get a value from lsattr -El {1} -a {2}", object, field);
+    // High-level logic here is:
+    //   * Check if there's an entry for our object's field attribute in CuAt (the device-specific
+    //   attribute entry).
+    //
+    //   * Else, check for the field attribute's default value in PdAt. We do this by first
+    //   figuring out the PdDv type from the CuDv entry for the object, then use our PdDv type
+    //   to query the field's default value in PdAt.
+    string query = (boost::format("name = %1% AND attribute = %2%") % object % field).str();
+    auto cdat_query = odm_class<CuAt>::open("CuAt").query(query);
+
+    // This is a more verbose way of saying that we only expect our query to have one element
+    auto cdat_ref = cdat_query.begin();
+    if (cdat_ref != cdat_query.end()) {
+      auto value = string((*cdat_ref).value);
+      if (value.empty()) {
+        LOG_DEBUG("Could not get a value from the ODM for {1}'s '{2}' attribute.", object, field);
+      }
+      return value;
+    }
+
+    // Get the PdDv type from the CuDv entry
+    query = (boost::format("name = %1%") % object).str();
+    auto cudv_query = odm_class<CuDv>::open("CuDv").query(query);
+    auto cudv_ref = cudv_query.begin();
+    if (cudv_ref == cudv_query.end()) {
+      LOG_DEBUG("Could not get a value from the ODM for {1}'s '{2}' attribute: There is no CuDv entry for {1}.", object, field);
+      return "";
+    }
+    auto pddv_type = (*cudv_ref).PdDvLn_Lvalue;
```

Comment:
should be `cudv_ref->PdDvLn_Lvalue`?

Reply:
Unfortunately no, GCC complains b/c `cudv_ref` is a different type. Here's the error.
```
/root/facter/lib/src/facts/aix/operating_system_resolver.cc: In function 'std::__cxx11::string getattr(std::__cxx11::string, std::__cxx11::string)':
/root/facter/lib/src/facts/aix/operating_system_resolver.cc:55:30: error: base operand of '->' has non-pointer type 'facter::util::aix::odm_class<CuDv>::iterator'
```

Reply:
we might just need to implement `operator->` for the iterator type?